kernel/object_validate.c

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <linker/linker-defs.h>
+#include <misc/__assert.h>
+#include <misc/printk.h>
+#include <init.h>
+#include <drivers/rand32.h>
+
+int initialized;
+
+struct object_metadata {
+	u8_t *start;
+	u8_t *end;
+	u32_t offset;
+	u32_t size;
+	u32_t xor_key;
+};
+
+#define OBJ_INFO(name) \
+	{ \
+	.start = &_##name##_list_start, \
+	.end = &_##name##_list_end, \
+	.size = sizeof(struct name), \
+	.offset = offsetof(struct name, obj), \
+	}
+
+#define OBJ_INFO_NOLIST(name) \
+	{ \
+	.start = NULL, \
+	.end = NULL, \
+	.size = sizeof(struct name), \
+	.offset = offsetof(struct name, obj), \
+	}
+
+/* For regions defined in linker/common-ram.ld */
+#define OBJ_BOUNDS(name) \
+	extern u8_t _##name##_list_start; \
+	extern u8_t _##name##_list_end
+
+OBJ_BOUNDS(k_alert);
+OBJ_BOUNDS(k_mem_slab);
+OBJ_BOUNDS(k_msgq);
+OBJ_BOUNDS(k_mutex);
+OBJ_BOUNDS(k_pipe);
+OBJ_BOUNDS(k_sem);
+OBJ_BOUNDS(k_stack);
+OBJ_BOUNDS(k_timer);
+OBJ_BOUNDS(k_work);
+
+static struct object_metadata ometa[K_OBJ_LAST] = {
+	[K_OBJ_ALERT] = OBJ_INFO(k_alert),
+	[K_OBJ_DELAYED_WORK] = OBJ_INFO_NOLIST(k_delayed_work),
+	[K_OBJ_MEM_SLAB] = OBJ_INFO(k_mem_slab),
+	[K_OBJ_MSGQ] = OBJ_INFO(k_msgq),
+	[K_OBJ_MUTEX] = OBJ_INFO(k_mutex),
+	[K_OBJ_PIPE] = OBJ_INFO(k_pipe),
+	[K_OBJ_SEM] = OBJ_INFO(k_sem),
+	[K_OBJ_STACK] = OBJ_INFO(k_stack),
+	[K_OBJ_THREAD] = OBJ_INFO_NOLIST(k_thread),
+	[K_OBJ_TIMER] = OBJ_INFO(k_timer),
+	[K_OBJ_WORK] = OBJ_INFO(k_work),
+	[K_OBJ_WORK_Q] = OBJ_INFO_NOLIST(k_work_q),
+};
+
+#define TO_OBJECT(ptr, type_enum) \
+	((u8_t *)ptr + ometa[type_enum].offset)
+
+#define FROM_OBJECT(ptr, type_enum) \
+	((u8_t *)ptr - ometa[type_enum].offset)
+
+static int init_k_obj_validate(struct device *unused)
+{
+	ARG_UNUSED(unused);
+	u8_t *pos;
+	enum k_objects i;
+
+	initialized = 1;
+
+	for (i = 0; i < K_OBJ_LAST; i++) {
+		struct object_metadata *o = &ometa[i];
+
+#if CONFIG_RANDOM_HAS_DRIVER
+		o->xor_key = sys_rand32_get();
```

Comment:
If we admit that this is totally insecure we might as well not have random numbers at all, and just use the loop counter (basically just the !CONFIG_RANDOM_HAS_DRIVER case) maybe with some math done on it to ensure not too close to other keys.

i guess let this patch be a prototype of the pointer-based API model. i'm going to prototype in another patch what a descriptor-based validation mechanism would look like, I think it will at least inform the discussion.